crates/control/src/services/builds_root.rs

```diff
@@ -0,0 +1,344 @@
+use crate::config;
+use crate::error::SubprocessError;
+use async_trait::async_trait;
+use rusqlite::{Connection, OpenFlags};
+use std::collections::BTreeMap;
+use std::fmt::Debug;
+use std::path::{Path, PathBuf};
+use std::sync::Arc;
+use std::time::{Duration, Instant};
+use tokio::sync::Mutex;
+
+mod gcs;
+
+/// Allows uploading builds to the builds root.
+#[derive(Debug, Clone)]
+pub struct PutBuilds(Arc<dyn BuildsRootService>);
```

Comment:
The idea was that our handlers would each only ever need one or the other. So an endpoint could just declare a `PutBuilds` parameter, without having to get both. I don't think it's important that it works that way, but I also don't really see a benefit to combining it.

Reply:
I think the main advantage of combining them would be to pre-cache items that are being added. Whenever a build is added to the builds root, I'd expect it to also end up in our fetch cache since we're very likely to turn around and need it immediately.